Description: Port hfi1 to kernel 4.9.0
Author: Brian T. Smith <bsmith@systemfabricworks.com>
Forwarded: not-needed
Last-Update: <2019-01-30>
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/Makefile
+++ b/Makefile
@@ -8,7 +8,8 @@
 #kbuild part of makefile
 
 CFLAGS_MODULE += -DUSE_PI_LED_ENABLE=1 -DIFS_DEB9
-obj-y := rdmavt/
+obj-y := rdmavt/ \
+	hfi1/
 
 else
 #normal makefile
--- a/hfi1/file_ops.c
+++ b/hfi1/file_ops.c
@@ -48,7 +48,8 @@
 #include <linux/cdev.h>
 #include <linux/vmalloc.h>
 #include <linux/io.h>
-#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_RH76) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3)
+#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_RH76) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3) \
+	&& !defined(IFS_DEB9)
 #include <linux/sched/mm.h>
 #else
 #include <linux/aio.h>
@@ -142,7 +143,7 @@
 static int ctxt_reset(struct hfi1_ctxtdata *uctxt);
 static int manage_rcvq(struct hfi1_ctxtdata *uctxt, u16 subctxt,
 		       unsigned long arg);
-#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_RH76) && !defined(IFS_RH77) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3)
+#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_RH76) && !defined(IFS_RH77) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3) && !defined(IFS_DEB9)
 static int vma_fault(struct vm_fault *vmf);
 #else
 static int vma_fault(struct vm_area_struct *vma, struct vm_fault *vmf);
@@ -693,7 +694,7 @@
  * Local (non-chip) user memory is not mapped right away but as it is
  * accessed by the user-level code.
  */
-#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_RH76) && !defined(IFS_RH77) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3)
+#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_RH76) && !defined(IFS_RH77) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3) && !defined(IFS_DEB9)
 static int vma_fault(struct vm_fault *vmf)
 #else
 static int vma_fault(struct vm_area_struct *vma, struct vm_fault *vmf)
--- a/hfi1/mad.c
+++ b/hfi1/mad.c
@@ -280,7 +280,8 @@
 	int ret = -EINVAL;
 
 	memset(&attr, 0, sizeof(attr));
-#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3)
+#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3) \
+	&& !defined(IFS_DEB9)
 	attr.type = ah->ibah.type;
 #endif
 	hfi1_update_sm_ah_attr(ibp, &attr, dlid);
@@ -297,7 +298,8 @@
 	struct rdma_ah_attr attr;
 	struct ib_ah *ah = ERR_PTR(-EINVAL);
 	struct rvt_qp *qp0;
-#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3)
+#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_RH75) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3) \
+	&& !defined(IFS_DEB9)
 	struct hfi1_pportdata *ppd = ppd_from_ibp(ibp);
 	struct hfi1_devdata *dd = dd_from_ppd(ppd);
 	u8 port_num = ppd->port;
--- a/hfi1/user_pages.c
+++ b/hfi1/user_pages.c
@@ -46,7 +46,8 @@
  */
 
 #include <linux/mm.h>
-#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3)
+#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3) \
+	&& !defined(IFS_DEB9)
 #include <linux/sched/signal.h>
 #else
 #include <linux/sched.h>
--- a/hfi1/verbs.c
+++ b/hfi1/verbs.c
@@ -1678,6 +1678,7 @@
 	RCU_INIT_POINTER(ibp->rvp.qp[0], NULL);
 	RCU_INIT_POINTER(ibp->rvp.qp[1], NULL);
 }
+
 #ifndef GET_DEV_FW_STR_HAS_LEN
 static void hfi1_get_dev_fw_str(struct ib_device *ibdev, char *str)
 {
@@ -1960,7 +1961,8 @@
 		ib_hfi1_sys_image_guid = ibdev->node_guid;
 	ibdev->owner = THIS_MODULE;
 	ibdev->phys_port_cnt = dd->num_pports;
-#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3)
+#if !defined(IFS_RH73) && !defined(IFS_RH74) && !defined(IFS_SLES12SP2) && !defined(IFS_SLES12SP3) \
+	&& !defined(IFS_DEB9)
 	ibdev->dev.parent = &dd->pcidev->dev;
 #else
 	ibdev->dma_device = &dd->pcidev->dev;
--- a/compat/compat.h
+++ b/compat/compat.h
@@ -51,10 +51,12 @@
 #define NEED_MM_HELPER_FUNCTIONS
 #define NEED_IB_HELPER_FUNCTIONS
 #define HAVE_NOSPEC_H
+#define GET_DEV_FW_STR_HAS_LEN
 
 #include "compat_common.h"
 
 #define IB_FW_VERSION_NAME_MAX			  ETHTOOL_FWVERS_LEN
+#define OPA_SM_CLASS_VERSION               	  0x80
 
 struct ib_ah *rdma_create_ah(struct ib_pd *pd, struct rdma_ah_attr *ah_attr);
 int rdma_destroy_ah(struct ib_ah *ah);
@@ -64,8 +66,6 @@
 
 #define ib_register_device(a, b, c)  ib_register_device((a), (c))
 #define rdma_set_device_sysfs_group(a, b)
-#define alloc_netdev_mqs(size, name, name_assign_type, setup, sdma, ctxts) \
-	alloc_netdev_mqs((size), (name), (setup), (sdma), (ctxts))
 #undef access_ok
 #define access_ok(addr, size) \
 	(likely(__range_not_ok(addr, size, user_addr_max()) == 0))
@@ -73,6 +73,4 @@
 
 void pcie_flr(struct pci_dev *dev);
 
-
-
 #endif //DEB9_COMPAT
--- a/hfi1/ipoib_main.c
+++ b/hfi1/ipoib_main.c
@@ -141,8 +141,13 @@
 	return priv->netdev_ops->ndo_stop(dev);
 }
 #if !defined(IFS_RH74) && !defined(IFS_SLES12SP2)
-static void hfi1_ipoib_dev_get_stats64(struct net_device *dev,
-				       struct rtnl_link_stats64 *storage)
+#if defined(IFS_DEB9)
+static struct rtnl_link_stats64*
+#else
+static void
+#endif
+hfi1_ipoib_dev_get_stats64(struct net_device *dev,
+			   struct rtnl_link_stats64 *storage)
 {
 	struct hfi1_ipoib_dev_priv *priv = hfi1_ipoib_priv(dev);
 	u64 rx_packets = 0ull;
@@ -180,6 +185,9 @@
 	storage->rx_bytes += rx_bytes;
 	storage->tx_packets += tx_packets;
 	storage->tx_bytes += tx_bytes;
+#if defined(IFS_DEB9)
+	return storage;
+#endif
 }
 #endif
 
--- a/hfi1/ipoib_tx.c
+++ b/hfi1/ipoib_tx.c
@@ -52,6 +52,7 @@
 
 #include <linux/log2.h>
 #include <linux/circ_buf.h>
+#include <linux/vmalloc.h>
 
 #include "sdma.h"
 #include "verbs.h"
--- a/hfi1/sdma.c
+++ b/hfi1/sdma.c
@@ -1482,9 +1482,20 @@
 		);
 		if (!sde->descq)
 			goto bail;
+#if defined(KVZALLOC_NODE)
 		sde->tx_ring =
 			kvzalloc_node(sizeof(struct sdma_txreq *) * descq_cnt,
 				      GFP_KERNEL, dd->node);
+#else
+		sde->tx_ring =
+			kcalloc(descq_cnt, sizeof(struct sdma_txreq *),
+				GFP_KERNEL);
+		if (!sde->tx_ring)
+			sde->tx_ring =
+				vzalloc(
+					sizeof(struct sdma_txreq *) *
+					descq_cnt);
+#endif
 		if (!sde->tx_ring)
 			goto bail;
 	}
